wp-content/themes/core/components/tabs/Tabs_Controller.php

```diff
@@ -0,0 +1,244 @@
+	public const TOGGLE_CLASSES     = 'toggle_classes';
+	public const TAB_BUTTON_CLASSES = 'tab_button_classes';
+	public const TAB_PANEL_CLASSES  = 'tab_panel_classes';
+	public const LAYOUT_HORIZONTAL  = 'horizontal';
+	public const LAYOUT_VERTICAL    = 'vertical';
+
+	/**
+	 * @var Tab_Model[] The collection of tabs to render. Each item should be a \Tribe\Project\Templates\Models\Tab object.
+	 */
+	private array $tabs;
+
+	private array $classes;
+	private array $attrs;
+	private string $layout;
+	private array $toggle_classes;
+	private array $tab_button_classes;
+	private array $tab_panel_classes;
+	private string $tablist_id;
+	private array $tab_buttons = [];
+	private array $tab_panels = [];
+
+	/**
+	 * Controller constructor.
+	 *
+	 * @param array $args
+	 */
+	public function __construct( array $args = [] ) {
+		$args = $this->parse_args( $args );
+
+		$this->tabs               = (array) $args[ self::TABS ];
+		$this->classes            = (array) $args[ self::CLASSES ];
+		$this->attrs              = (array) $args[ self::ATTRS ];
+		$this->layout             = (string) $args[ self::LAYOUT ];
+		$this->toggle_classes     = (array) $args[ self::TOGGLE_CLASSES ];
+		$this->tab_button_classes = (array) $args[ self::TAB_BUTTON_CLASSES ];
+		$this->tab_panel_classes  = (array) $args[ self::TAB_PANEL_CLASSES ];
+		$this->tablist_id         = uniqid( 'c-tabs__tablist--' );
+
+		$this->init_tabs();
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function defaults(): array {
+		return [
+			self::TABS               => [],
+			self::CLASSES            => [],
+			self::ATTRS              => [],
+			self::LAYOUT             => self::LAYOUT_HORIZONTAL,
+			self::TOGGLE_CLASSES     => [],
+			self::TAB_BUTTON_CLASSES => [],
+			self::TAB_PANEL_CLASSES  => [ 's-sink', 't-sink' ],
+		];
+	}
+
+	/**
+	 * @return array
+	 */
+	protected function required(): array {
+		return [
+			self::CLASSES            => [ 'c-tabs' ],
+			self::ATTRS              => [ 'data-js' => 'c-tabs' ],
+			self::TOGGLE_CLASSES     => [ 'c-tabs__tablist-toggle' ],
+			self::TAB_BUTTON_CLASSES => [ 'c-tabs__tab' ],
+			self::TAB_PANEL_CLASSES  => [ 'c-tabs__tabpanel' ],
+		];
+	}
+
+	/**
+	 * Loop through the tabs provided setup the tab button and tab panel components arguments for each.
+	 */
+	private function init_tabs() {
+		foreach ( $this->tabs as $index => $tab ) {
+			$tab_id              = uniqid();
+			$this->tab_buttons[] = $this->get_tab_button_args( $tab, $tab_id, $index );
+			$this->tab_panels[]  = $this->get_tab_panel_args( $tab, $tab_id, $index );
+		}
+	}
+
+	/**
+	 * Return arguments for an individual tab list "tab" button.
+	 *
+	 * @param Tab_Model $tab
+	 * @param string    $tab_id
+	 * @param int       $index
+	 *
+	 * @return array
+	 */
+	private function get_tab_button_args( Tab_Model $tab, string $tab_id, int $index ): array {
+		$args = [
+			'content' => $tab->label ?: sprintf( __( 'Tab %d', 'tribe' ), $index + 1 ),
```

Comment:
Escape $tab->label? I don't think it's escaped anywhere else down the line.

Reply:
I think maybe we need to do a pass on escaping in general.

Reply:
What difficult about this is I can see times where we'll pass arbitrary html into a button (a span, an icon SVG or <i>, etc).  So, what should we allow here? Maybe just use `wp_kses_post()`?